lib/bolt/cli.rb

```diff
@@ -564,19 +564,121 @@ def generate_types
       0
     end
 
+    # Initializes a specified directory as a Bolt project and installs any modules
+    # specified by the user, along with their dependencies
     def initialize_project
-      path = File.expand_path(options[:object] || Dir.pwd)
-      FileUtils.mkdir_p(path)
-      ok = FileUtils.touch(File.join(path, 'bolt.yaml'))
+      boltdir    = Pathname.new(File.expand_path(options[:object] || Dir.pwd))
+      config     = boltdir + 'bolt.yaml'
+      puppetfile = boltdir + 'Puppetfile'
+      modulepath = [boltdir + 'modules']
+      plural     = options[:modules]&.count == 1 ? '' : 's'
+
+      # If modules were specified, first check if there is already a Puppetfile at the project
+      # directory, erroring if there is. If there is no Puppetfile, generate the Puppetfile
+      # content by resolving the specified modules and all their dependencies.
+      # We generate the Puppetfile first so that any errors in resolving modules and their
+      # dependencies are caught early and do not create a project directory.
+      if options[:modules]
+        if puppetfile.exist?
+          raise Bolt::CLIError,
+                "Found existing Puppetfile at #{puppetfile}, unable to initialize project with "\
+                "module#{plural} #{options[:modules].join(', ')}"
+        else
+          puppetfile_specs = resolve_puppetfile_specs
+        end
+      end
 
-      result = if ok
-                 "Successfully created Bolt project directory at #{path}"
-               else
-                 "Could not create Bolt project directory at #{path}"
-               end
-      outputter.print_message result
+      # Warn the user if the project directory already exists. We don't error here since users
+      # might not have installed any modules yet.
+      if config.exist?
+        @logger.warn "Found existing project directory at #{boltdir}"
+      end
 
-      ok ? 0 : 1
+      # Create the project directory
+      FileUtils.mkdir_p(boltdir)
+
+      # Bless the project directory as a boltdir
+      if FileUtils.touch(config)
+        outputter.print_message "Successfully created Bolt project at #{boltdir}"
+      else
+        raise Bolt::FileError.new("Could not create Bolt project directory at #{boltdir}", nil)
+      end
+
+      # Write the generated Puppetfile to the fancy new boltdir
+      if puppetfile_specs
+        if File.write(puppetfile, puppetfile_specs.join("\n"))
```

Comment:
It looks like `File.write` returns an integer, so this will never be false. I expect that if it writing the file fails, it will raise an error. It's probably fine to just let that fail, since we've done some preliminary validation already (and so failing at this point is indeed exceptional).

Reply:
Not sure if this is an issue, but File.write will use CRLF line endings on Windows.  Not sure if you actually expect LF on all platforms?